cpp/src/io/parquet/parquet.hpp

```diff
@@ -346,6 +345,270 @@ class CompactProtocolReader {
+  template <typename T>
+  friend class ParquetFieldEnum;
+  template <typename T>
+  friend class ParquetFieldEnumListFunctor;
+  friend class ParquetFieldStringList;
+  friend class ParquetFieldStructBlob;
+};
+
+/**
+ * @brief Functor to set value to 32 bit integer read from CompactProtocolReader
+ *
+ * @return True if field type is not int32
+ */
+class ParquetFieldInt32 {
+  int field_val;
+  int32_t &val;
+
+ public:
+  ParquetFieldInt32(int f, int32_t &v) : field_val(f), val(v) {}
+
+  inline bool operator()(CompactProtocolReader *cpr, int field_type)
+  {
+    val = cpr->get_i32();
+    return (field_type != ST_FLD_I32);
+  }
+
+  int &field(void) { return field_val; }
+};
+
+/**
+ * @brief Functor to set value to 64 bit integer read from CompactProtocolReader
+ *
+ * @return True if field type is not int32 or int64
+ */
+class ParquetFieldInt64 {
+  int field_val;
+  int64_t &val;
+
+ public:
+  ParquetFieldInt64(int f, int64_t &v) : field_val(f), val(v) {}
+
+  inline bool operator()(CompactProtocolReader *cpr, int field_type)
+  {
+    val = cpr->get_i64();
+    return (field_type < ST_FLD_I16 || field_type > ST_FLD_I64);
+  }
+
+  int &field(void) { return field_val; }
+};
+
+/**
+ * @brief Functor to read a vector of structures from CompactProtocolReader
+ *
+ * @return True if field types mismatch or if the process of reading a
+ * struct fails
+ */
+template <typename T>
+class ParquetFieldStructListFunctor {
+  int field_val;
+  std::vector<T> &val;
+
+ public:
+  ParquetFieldStructListFunctor(int f, std::vector<T> &v) : field_val(f), val(v) {}
+
+  inline bool operator()(CompactProtocolReader *cpr, int field_type)
+  {
+    if (field_type != ST_FLD_LIST) return true;
+
+    int current_byte = cpr->getb();
+    if ((current_byte & 0xf) != ST_FLD_STRUCT) return true;
+    int n = current_byte >> 4;
+    if (n == 0xf) n = cpr->get_u32();
+    val.resize(n);
+    for (int32_t i = 0; i < n; i++) {
+      if (!(cpr->read(&val[i]))) { return true; }
+    }
+
+    return false;
+  }
+
+  int &field(void) { return field_val; }
+};
+
+template <typename T>
+ParquetFieldStructListFunctor<T> ParquetFieldStructList(int f, std::vector<T> &v)
+{
+  return ParquetFieldStructListFunctor<T>(f, v);
+}
+
+/**
+ * @brief Functor to read a string from CompactProtocolReader
+ *
+ * @return True if field type mismatches or if size of string exceeds bounds
+ * of the CompactProtocolReader
+ */
+class ParquetFieldString {
+  int field_val;
+  std::string &val;
+
+ public:
+  ParquetFieldString(int f, std::string &v) : field_val(f), val(v) {}
+
+  inline bool operator()(CompactProtocolReader *cpr, int field_type)
+  {
+    if (field_type != ST_FLD_BINARY) return true;
+    uint32_t n = cpr->get_u32();
+    if (n < (size_t)(cpr->m_end - cpr->m_cur)) {
+      val.assign((const char *)cpr->m_cur, n);
+      cpr->m_cur += n;
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  int &field(void) { return field_val; }
```

Comment:
You can return by value in `field`getters. Also, no need to put `void` here. This goes for all `field` implementations, AFAICT.
```suggestion
  int field() { return field_val; }
```